File: ios/Classes/CallHandlers/CountersCallHandler.swift
//
//  CountersCallHandler.swift
//  Flutter-SDK
//
//  Created by Andrii Bodnar on 4/30/19.
//  Copyright © 2019 Andrii Bodnar. All rights reserved.
//

import Foundation
import Flutter
import Backendless

class CountersCallHandler: FlutterCallHandlerProtocol {
    
    // MARK: -
    // MARK: - Constants
    private enum Methods {
        static let addAndGet = "Backendless.Counters.addAndGet"
        static let compareAndSet = "Backendless.Counters.compareAndSet"
        static let decrementAndGet = "Backendless.Counters.decrementAndGet"
        static let get = "Backendless.Counters.get"
        static let getAndAdd = "Backendless.Counters.getAndAdd"
        static let getAndDecrement = "Backendless.Counters.getAndDecrement"
        static let getAndIncrement = "Backendless.Counters.getAndIncrement"
        static let incrementAndGet = "Backendless.Counters.incrementAndGet"
        static let reset = "Backendless.Counters.reset"
    }
    
    private enum Args {
        static let counterName = "counterName"
        static let value = "value"
        static let expected = "expected"
        static let updated = "updated"
    }
    
    // MARK: -
    // MARK: - Counters Reference
    private let counters = SwiftBackendlessSdkPlugin.backendless.counters
    
    // MARK: -
    // MARK: - Route Flutter Call
    func routeFlutterCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        guard
            let arguments: [String: Any] = call.arguments.flatMap(cast),
            let counterName: String = arguments[Args.counterName].flatMap(cast)
        else {
            result(FlutterError.noRequiredArguments)
            
            return
        }
        
        switch call.method {
        case Methods.addAndGet:
            self.addAndGet(counterName, arguments, result)
        case Methods.compareAndSet:
            self.compareAndSet(counterName, arguments, result)
        case Methods.decrementAndGet:
            self.decrementAndGet(counterName, result)
        case Methods.get:
            self.get(counterName, result)
        case Methods.getAndAdd:
            self.getAndAdd(counterName, arguments, result)
        case Methods.getAndDecrement:
            self.getAndDecrement(counterName, result)
        case Methods.getAndIncrement:
            self.getAndIncrement(counterName, result)
        case Methods.incrementAndGet:
            self.incrementAndGet(counterName, result)
        case Methods.reset:
            self.reset(counterName, result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: -
    // MARK: - AddAndGet
    private func addAndGet(_ counterName: String, _ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard let value: Int = arguments[Args.value].flatMap(cast) else {
            result(FlutterError.noRequiredArguments)
            
            return
        }
        
        counters.addAndGet(counterName: counterName, value: value,
            responseHandler: {
                result($0)
            },
            errorHandler: {
                result(FlutterError($0))
            })
    }
    
    // MARK: -
    // MARK: - CompareAndSet
    private func compareAndSet(_ counterName: String, _ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard
            let expected: Int = arguments[Args.expected].flatMap(cast),
            let updated: Int = arguments[Args.updated].flatMap(cast)
        else {
            result(FlutterError.noRequiredArguments)
            
            return
        }
        
        counters.compareAndSet(counterName: counterName, expected: expected, updated: updated,
            responseHandler: {
                result($0)
            },
            errorHandler: {
                result(FlutterError($0))
            })
    }
    
    // MARK: -
    // MARK: - DecrementAndGet
    func decrementAndGet(_ counterName: String, _ result: @escaping FlutterResult) {
        counters.decrementAndGet(counterName: counterName,
            responseHandler: {
                result($0)
            },
            errorHandler: {
                result(FlutterError($0))
            })
    }
    
    // MARK: -
    // MARK: - Get
    private func get(_ counterName: String, _ result: @escaping FlutterResult) {
        counters.get(counterName: counterName,
            responseHandler: {
                result($0)
            },
            errorHandler: {
                result(FlutterError($0))
            })
    }
    
    // MARK: -
    // MARK: - GetAndAdd
    private func getAndAdd(_ counterName: String, _ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard let value: Int = arguments[Args.value].flatMap(cast) else {
            result(FlutterError.noRequiredArguments)
            
            return
        }
        
        counters.getAndAdd(counterName: counterName, value: value,
            responseHandler: {
                result($0)
            },
            errorHandler: {
                result(FlutterError($0))
            })
    }
    
    // MARK: -
    // MARK: - GetAndDecrement
    private func getAndDecrement(_ counterName: String, _ result: @escaping FlutterResult) {
        counters.getAndDecrement(counterName: counterName,
            responseHandler: {
                result($0)
            },
            errorHandler: {
                result(FlutterError($0))
            })
    }
    
    // MARK: -
    // MARK: - GetAndIncrement
    private func getAndIncrement(_ counterName: String, _ result: @escaping FlutterResult) {
        counters.getAndIncrement(counterName: counterName,
            responseHandler: {
                result($0)
            },
            errorHandler: {
                result(FlutterError($0))
            })
    }
    
    // MARK: -
    // MARK: - IncrementAndGet
    private func incrementAndGet(_ counterName: String, _ result: @escaping FlutterResult) {
        counters.incrementAndGet(counterName: counterName,
            responseHandler: {
                result($0)
            },
            errorHandler: {
                result(FlutterError($0))
            })
    }
    
    // MARK: -
    // MARK: - Reset
    private func reset(_ counterName: String, _ result: @escaping FlutterResult) {
        counters.reset(counterName: counterName,
            responseHandler: {
                result(nil)
            },
            errorHandler: {
                result(FlutterError($0))
            })
    }
}

File: android/src/main/java/com/backendless/backendless_sdk/call_handlers/CountersCallHandler.java
package com.backendless.backendless_sdk.call_handlers;

import com.backendless.Backendless;
import com.backendless.BackendlessInjector;
import com.backendless.async.callback.AsyncCallback;
import com.backendless.backendless_sdk.utils.FlutterCallback;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class CountersCallHandler implements MethodChannel.MethodCallHandler {

    public CountersCallHandler() {

    }

    @Override
    public void onMethodCall(MethodCall call, MethodChannel.Result result) {
        String counterName = call.argument("counterName");
        switch (call.method) {
            case ("Backendless.Counters.addAndGet"):
                addAndGet(counterName, call, result);
                break;
            case ("Backendless.Counters.compareAndSet"):
                compareAndSet(counterName, call, result);
                break;
            case ("Backendless.Counters.decrementAndGet"):
                decrementAndGet(counterName, result);
                break;
            case ("Backendless.Counters.get"):
                get(counterName, result);
                break;
            case ("Backendless.Counters.getAndAdd"):
                getAndAdd(counterName, call, result);
                break;
            case ("Backendless.Counters.getAndDecrement"):
                getAndDecrement(counterName, result);
                break;
            case ("Backendless.Counters.getAndIncrement"):
                getAndIncrement(counterName, result);
                break;
            case ("Backendless.Counters.incrementAndGet"):
                incrementAndGet(counterName, result);
                break;
            case ("Backendless.Counters.reset"):
                reset(counterName, result);
                break;
            default:
                result.notImplemented();
        }
    }

    @Deprecated
    private void addAndGet(String counterName, MethodCall call, final MethodChannel.Result result) {
        Number value = call.argument("value");
        Backendless.Counters.addAndGet(counterName, value,  new FlutterAtomicCallback(result));
    }

    @Deprecated
    private void compareAndSet(String counterName, MethodCall call, MethodChannel.Result result) {
        Number expected = call.argument("expected");
        Number updated = call.argument("updated");
        Backendless.Counters.compareAndSet(counterName, expected, updated, new FlutterCallback<Boolean>(result));
    }

    @Deprecated
    private void decrementAndGet(String counterName, MethodChannel.Result result) {
        Backendless.Counters.decrementAndGet(counterName, new FlutterAtomicCallback(result));
    }

    @Deprecated
    private void get(String counterName, MethodChannel.Result result) {
        Backendless.Counters.get(counterName, new FlutterAtomicCallback(result));
    }

    @Deprecated
    private void getAndAdd(String counterName, MethodCall call, MethodChannel.Result result) {
        Number value = call.argument("value");
        Backendless.Counters.getAndAdd(counterName, value, new FlutterAtomicCallback(result));
    }

    @Deprecated
    private void getAndDecrement(String counterName, MethodChannel.Result result) {
        Backendless.Counters.getAndDecrement(counterName, new FlutterAtomicCallback(result));
    }

    @Deprecated
    private void getAndIncrement(String counterName, MethodChannel.Result result) {
        Backendless.Counters.getAndIncrement(counterName, new FlutterAtomicCallback(result));
    }

    @Deprecated
    private void incrementAndGet(String counterName, MethodChannel.Result result) {
        Backendless.Counters.incrementAndGet(counterName, new FlutterAtomicCallback(result));
    }

    @Deprecated
    private void reset(String counterName, MethodChannel.Result result) {
        Backendless.Counters.reset(counterName, new FlutterCallback<>(result));
    }

//    Workaround to bypass ReflectionUtil.getCallbackGenericType call in AtomicCallback
    class FlutterAtomicCallback extends FlutterCallback<Long> implements AsyncCallback<Long> {
        public FlutterAtomicCallback(MethodChannel.Result result) { super(result); }
    }
}
